src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -276,26 +276,52 @@ internal static class StreamHelper
 
         #region Static Methods
 
-        internal static void WriteToStream(Stream input, Stream output, PSCmdlet cmdlet, CancellationToken cancellationToken)
+        internal static void WriteToStream(Stream input, Stream output, WebRequestPSCmdlet cmdlet, CancellationToken cancellationToken)
         {
             if (cmdlet == null)
             {
                 throw new ArgumentNullException(nameof(cmdlet));
             }
 
-            Task copyTask = input.CopyToAsync(output, cancellationToken);
-
             ProgressRecord record = new(
                 ActivityId,
                 WebCmdletStrings.WriteRequestProgressActivity,
                 WebCmdletStrings.WriteRequestProgressStatus);
 
+            // If the client lost target service on ~1 min due to a lost of network connectivity
+            // the CopyToAsync() task will be never completed and cycle below will be infinite.
+            // Short explaination:
+            // - CopyToAsync() uses ReadAsync() in cycle until ReadAsync() read 0 bytes (it is EOF).
+            // - ReadAsync() read from network Socket.
+            // - If network lost connectivity over 1 min the Socket lost connect to target service
+            //   but OS never close the Socket (tested on Windows).
+            // - Then Socket never return an information, ReadAsync() will be infinitely blocked on the Socket
+            //   and CopyToAsync() task will be never completed.
+            //
+            // Since cancelation token in CopyToAsync() applies to whole CopyToAsync()
+            // and it is not restarted for very internal ReadAsync()
+            // we have to use a workaround to cancel the ReadAsync() which is blocked on Socket longer Timeout.
+            // The workaround is to reset the cancelation timer
+            // while the length of the output file is constantly increasing.
+            long previousLength = 0;
+            var timeout = cmdlet.TimeoutSec == 0 ? Timeout.Infinite : cmdlet.TimeoutSec * 1000;
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            cts.CancelAfter(timeout);
+            Task copyTask = input.CopyToAsync(output, cts.Token);
+
             try
             {
-                while (!copyTask.Wait(1000, cancellationToken))
+                while (!copyTask.Wait(1000, cts.Token))
                 {
                     record.StatusDescription = StringUtil.Format(WebCmdletStrings.WriteRequestProgressStatus, output.Position);
                     cmdlet.WriteProgress(record);
+
+                    if (previousLength != output.Length)
+                    {
+                        // Reset cancelation timer only if we get an information from network on the interation.
```

Comment:
This seems like a dotNet bug, where cancelation timer is ineffective during stream copy task.  Does it make sense to submit an issue report about this?
Also, the timeout time is now only for when there is no network connectivity.  User expectation is that the timeout time is for the entire process including both connected and non-connected times.  We would need to update the documentation to make clear what the `-TimeOutSec` parameter does.

```suggestion
                        // Reset cancelation timer while information continues to flow from network.
                        // Cancelation timer applies only during no network connectivity.
```

Reply:
> This seems like a dotNet bug, where cancelation timer is ineffective during stream copy task. Does it make sense to submit an issue report about this?

This works for _whole_ CopyToAsync but it is not that we want since full download can take unpredictable time. We need only monitor every internal request to target service for hang - the CopyToAsync has internal cycle where ReadAsync() is used. It is just the ReadAsync hangs and we need reset the timer for every iteration in the cycle.

That you assumed it was a bug in the .Net Runtime is an interesting psychological experiment. :-) That was my assumption too, and I opened a [discussion ](https://github.com/dotnet/runtime/issues/74568 )there a few days ago. Only the .Net team offers workarounds and doesn't agree that HttpClient.Timeout should apply to operations on streams _it creates itself_, even though it works with all its other operations.
You could share your expectations there too if you want.

> We would need to update the documentation to make clear what the -TimeOutSec parameter does.

Both .Net and PowerShell docs should be updated for the Timeout parameter - that it applies to every request to target service and protects from too long (infinite) network operations.